### DISCUSSION TITLE
How do I configure my track point to middle click on middle button press?

### DISCUSSION THREAD

**Discussion by lilydjwg** [hyprwm/Hyprland#2490] [Q&A] [answered]
I have a ThinkPad laptop and it comes with a track point device. Pressing the middle button doesn't send any events but make moving mouse with the red dot scroll. I want it to middle click instead but I couldn't find the correct config option.

**Comment by cwegener** [comment]
~~That's interesting. I only just noticed this problem as well. It's not a specific problem for hyprland though, at least dwl exhibits the same behaviour.~~

~~I wonder what happens on GNOME's Wayland compositor.~~

~~I think the relevant feature in libinput is this feature~~: https://wayland.freedesktop.org/libinput/doc/latest/scrolling.html#on-button-scrolling

Edit: see my comments below. The function works just fine for me (on a Thinkpad T14 Gen1) in both dwl and Hyprland

**Reply by lilydjwg** [reply]
Wayfire (at least my fork from an earlier version; it also uses wlroots) does middle clicks by default.

**Reply by cwegener** [reply]
Wayfire did use libweston instead of wlroots at some point in the past ... https://wayfire.org/2019/02/24/X11-WLC-Libweston-Wlroots.html

**Reply by cwegener** [reply]
Ah. Sorry.

Ignore my initial comment.

Middle mouse button + trackpoint works perfectly for scrolling in dwl ....

**Reply by cwegener** [reply]
I just tested the same functionality on Hyprland and the middle mouse button + trackpoint function works just fine in Hyprland as well for me.

**Reply by lilydjwg** [reply]
Yes, it works fine to scroll, but I want it to middle click, because for most of time two-finger scroll works good enough for me, but I need middle clicks very often (paste selection, open background links, close tabs and windows, drag canvas, etc).

**Reply by cwegener** [reply]
Ah.

Ok.

That's interesting.

I wonder if that has something to do with the fact that the middle button on the thinkpad does not generate a 'pressed' event when it's pressed, but it only generates the 'pressed' event when the button is released (together with the 'released' event)

The left and right buttons don't behave like that. left and right buttons generate the 'pressed' event immediately when the buttons are pressed.

(I have observed the above using both, `wev` and `libinput debug-events`)

**Reply by lilydjwg** [reply]
`evtest` shows that there are separate events for press and release. So it's libinput taking them together.

**Reply by cwegener** [reply]
For comparison. I just tested with an external USB mouse, and the middle button press event is immediately registered in libinput as shown in `wev` and `libinput debug-events`

And as a result, pasting from the primary clipboard works in this case (at least in `dwl` but I suspect `hyprland` will be the same .. I'll do a quick hyprland test with an external mouse later)

So, based on your finding about `evtest` and my results with an external mouse, I think that the libinput link I pasted in the beginning might actually be related after all https://wayland.freedesktop.org/libinput/doc/latest/scrolling.html#on-button-scrolling

**Reply by cwegener** [reply]
I also confirmed that an external mouse works as expected in `hyprland` as well, the same way it does in `dwl`. The middle mouse button on the external mouse functions as a 'paste' button as expected.

So, the issue does appear to be related to the 'on button scrolling' feature in libinput.

**Reply by cwegener** [reply]
I'm just reading through this ... https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/4334

**Reply by lilydjwg** [reply]
Yes this relates to the on button scrolling feature. Usually there are options for user to tweak.

**Reply by cwegener** [reply]
According to the GNOME issue linked above, the problem did exist in GNOME's wayland shell (i.e. mutter) for a while, and then the problem disappeared in GNOME 42.4 without the issue author knowing what changed. So, GNOME doesn't seem like a good option to use as an example for comparison.

Do you know of any Wayland DE or WM that has this configuration option?

**Reply by lilydjwg** [reply, marked as answer]
`scroll_method` seems to be the option controlling this. Currently I don't reproduce this issue and it behaves in a mixed way, likely because I've run Wayfire and it has changed the option. This other possibility is I misunderstood the issue at the time opening this discussion.
